File: 최승훈/3주차/2343.py
import sys

n, m = map(int, sys.stdin.readline().split())
data = list(map(int, sys.stdin.readline().split()))

# 순서를 유지해야 하므로 정렬을 사용하면 안된다.
# 모두 담을 수 있어야 하므로 start = max(data)
# 블루레이의 크기 중 최소이므로 sum(data)보다는 작다.
num = sum(data)
start = max(data)
end = num
result = 1e9 + 7

while start <= end:
    # mid는 블루레이 크기(size)
    mid = (start + end) // 2
    # count는 블루레이 개수 , length는 블루레이 길이
    count, length = 1, 0
    for i in range(len(data)):
        # length는 지금까지 블루레이 크기의 합, data[i] 다음 블루레이 크기
        if length + data[i] <= mid:
            length += data[i]
        else:
            # mid 보다 커지면 현재 data[i] -> length 개수 1개 늘려준다.
            length = data[i]
            count += 1
    # 개수가 m보다 작거나 같으면 이분 탐색 실행
    if count <= m:
        end = mid - 1
        # mid 중에 제일 작은 값 저장
        result = min(result, mid)
    else:
        start = mid + 1
print(result)

File: 김영진/3주차/2343.c
#include<stdio.h>

int main() {
	int n, m;
	scanf("%d %d", &n, &m);
	int time[100001] = { 0 };
	int max = 0, sum = 0;
	for (int i = 0; i < n; i++) {
		scanf("%d", &time[i]);
		sum += time[i];
		if (time[i] > max) {
			max = time[i];
		}
	}
	int start = max;
	int end = sum;
	int mid;
	int j, count;
	int min_time = 1000000000;
	int ex = 0;
	while (start <= end) {
		mid = (start + end) / 2;
		j = 0;
		int bd[100000] = { 0 };
		int bd_max = 0;
		count = 0;
		for (int i = 0; i < m; i++) {
			while (bd[i] <= mid) {
				if (time[j] == 0) {
					break;
				}
				bd[i] += time[j];
				j++;
			}
			if (time[j] != 0 || bd[i] > mid) {
				bd[i] -= time[--j];
			}
			if (bd[i] > bd_max) {
				bd_max = bd[i];
			}
			count++;
			if (time[j] == 0) {
				break;
			}
		}
		if (count < m) {
			if (bd_max < min_time) {
				min_time = bd_max;
			}
			end = mid - 1;
		}
		else if (time[j] != 0) {
			start = mid + 1;
		}
		else if (count == m) {
			if (bd_max < min_time) {
				min_time = bd_max;
				ex++;
			}
			if (start == mid) {
				break;
			}
			end = mid - 1;
		}
	}
	printf("%d", min_time);

	return 0;
}
